# 📊 Data Preprocessing

The primary objective of this notebook is to **preprocess** the raw Bitcoin data. This step is essential to ensure the dataset is structured performing Exploratory Data Analysis (EDA) and Feature Engineering.

## 1. Data Acquisition

We acquire historical hourly data for Bitcoin  from 2018 to the present. To achieve this, we utilize the custom script `src/get_ccxt_data.py`, which interfaces with the **Binance API** via the `ccxt` library to fetch the complete OHLCV history without gaps.

In [81]:
import pandas as pd
import numpy as np
# Importing the downloaded data from the data carpeta
df = pd.read_csv('../data/btcusdt_1h_2018.csv')
df

Date      Open      High       Low     Close      Volume
0      2018-01-01 00:00:00  13715.65  13715.65  13400.01  13529.01  443.356199
1      2018-01-01 01:00:00  13528.99  13595.89  13155.38  13203.06  383.697006
2      2018-01-01 02:00:00  13203.00  13418.43  13200.00  13330.18  429.064572
3      2018-01-01 03:00:00  13330.26  13611.27  13290.00  13410.03  420.087030
4      2018-01-01 04:00:00  13434.98  13623.29  13322.15  13601.01  340.807329
...                    ...       ...       ...       ...       ...         ...
69098  2025-11-24 04:00:00  87460.03  87600.00  86450.00  86738.37  878.958090
69099  2025-11-24 05:00:00  86738.36  87837.57  86675.82  87468.29  753.252170
69100  2025-11-24 06:00:00  87468.28  87518.06  86819.98  86910.78  801.385800
69101  2025-11-24 07:00:00  86910.77  87350.00  86700.00  87050.39  842.743540
69102  2025-11-24 08:00:00  87050.39  87050.39  86724.78  86783.86  509.831590

[69103 rows x 6 columns]

As we can see above, there are 69103 rows, which makes a big amount of data. 

## 2. Data Loading & Inspection

We load the raw CSV file into a Pandas DataFrame to inspect its structure and data types.

In [82]:
# Data info
df.info()
# Statistical summary
df[['Open', 'High', 'Low', 'Close', 'Volume']].describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69103 entries, 0 to 69102
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    69103 non-null  object 
 1   Open    69103 non-null  float64
 2   High    69103 non-null  float64
 3   Low     69103 non-null  float64
 4   Close   69103 non-null  float64
 5   Volume  69103 non-null  float64
dtypes: float64(5), object(1)
memory usage: 3.2+ MB


Open           High            Low          Close  \
count   69103.000000   69103.000000   69103.000000   69103.000000   
mean    36600.063961   36748.968598   36443.602233   36601.120899   
std     31871.291714   31969.792975   31769.212601   31871.746127   
min      3172.620000    3184.750000    3156.260000    3172.050000   
25%      9391.695000    9431.695000    9350.415000    9391.375000   
50%     27025.890000   27100.010000   26925.800000   27028.960000   
75%     56250.825000   56550.920000   55826.980000   56250.820000   
max    126011.180000  126199.630000  125252.740000  126011.180000   

              Volume  
count   69103.000000  
mean     2749.358070  
std      3993.867442  
min         0.000000  
25%       855.343804  
50%      1505.258660  
75%      2888.109051  
max    137207.188600

There are no null values, so there is no matter to worry about them.

## 3. Formatting and Type Conversion

The initial inspection reveals that the `Date` column is currently stored as a string (`object`). For time-series analysis, it is crucial to convert this column to `datetime64[ns]` objects and set it as the DataFrame index. 

In [83]:
# Date to datetime
df['Date'] = pd.to_datetime(df['Date'])
df.set_index('Date', inplace=True)
df.sort_index(inplace=True)
df.head()

Open      High       Low     Close      Volume
Date                                                                   
2018-01-01 00:00:00  13715.65  13715.65  13400.01  13529.01  443.356199
2018-01-01 01:00:00  13528.99  13595.89  13155.38  13203.06  383.697006
2018-01-01 02:00:00  13203.00  13418.43  13200.00  13330.18  429.064572
2018-01-01 03:00:00  13330.26  13611.27  13290.00  13410.03  420.087030
2018-01-01 04:00:00  13434.98  13623.29  13322.15  13601.01  340.807329

## 4. Saving the Processed Dataset

Finally, we save the cleaned and preprocessed DataFrame to a new CSV file. 

In [87]:
# We dave the df as a csv file
df.to_csv('../data/btcusdt_1h_2018_preprocessed.csv')   

## 🔄 Processing Additional Timeframes (1D & 5m)

We repeat the cleaning pipeline (Type Conversion -> Indexing -> Sorting) for the Daily (`1d`) and 5-Minute (`5m`) datasets.

In [88]:
# With 1 day timeframe
df_1d = pd.read_csv('../data/btcusdt_1d_2018.csv')
df_1d['Date'] = pd.to_datetime(df_1d['Date'])
df_1d.set_index('Date', inplace=True)
df_1d.sort_index(inplace=True)
df_1d.to_csv('../data/btcusdt_1d_2018_preprocessed.csv')
# With 5 minutes timeframe    
df_5m = pd.read_csv('../data/btcusdt_5m_2018.csv')
df_5m['Date'] = pd.to_datetime(df_5m['Date'])
df_5m.set_index('Date', inplace=True)
df_5m.sort_index(inplace=True)
df_5m.to_csv('../data/btcusdt_5m_2018_preprocessed.csv')    